index.ts

```diff
@@ -0,0 +1,425 @@
+      }
+    );
+
+    db.materialCategories = updatedMaterialCategories;
+
+    console.log("DELETE /categories/{id} DB:", db);
+
+    res.send("Category deleted successfully");
+  } catch (error) {
+    console.log(error);
+    res.send("Error in delete one category");
+  }
+});
+
+// Materials endpoints
+// Create one material
+app.post("/materials", (req: Request<{}, {}, MaterialDef, {}>, res) => {
+  try {
+    const body: unknown = req.body;
+    const parsedBody = materialDefSchema.parse(body);
+    const id = crypto.randomUUID();
+    const { categoryIds, ...materialBase } = parsedBody;
+
+    const material: Material = {
+      ...materialBase,
+      id,
+      createdAt: Date.now(),
+      updatedAt: Date.now(),
+    };
+
+    db.materials.push(material);
+
+    for (const categoryId of categoryIds) {
+      const materialCategory: MaterialCategory = { categoryId, materialId: id };
+      db.materialCategories.push(materialCategory);
+    }
+
+    console.log("POST /materials DB:", db);
+
+    res.send(material);
+  } catch (error) {
+    console.error(error);
+
+    if (error instanceof ZodError) {
+      res.status(422).send("Error creating material");
+      return;
+    }
+
+    res.status(500).send("Unknown error");
+  }
+});
+
+// Get all materials
+app.get("/materials", (req, res) => {
+  try {
+    const materials = db.materials;
+    res.send(materials);
+  } catch (error) {
+    console.log(error);
+    res.send("Error in get all materials");
+  }
+});
+
+// Get one material
+app.get("/materials/:id", (req, res) => {
+  try {
+    const { id } = req.params;
+    const foundMaterial = db.materials.find((material) => material.id === id);
+
+    if (foundMaterial) {
+      res.send(foundMaterial);
+      return;
+    }
+
+    res.send(`Material with provided ${id} is not found`);
+  } catch (error) {
+    console.log(error);
+    res.send("Error in get one material");
+  }
+});
+
+// Get all materials by one category
+app.get("/materials/category/:categoryId", (req, res) => {
+  try {
+    const { categoryId } = req.params;
+
+    const filteredMaterialIdsByCategory = db.materialCategories
+      .filter((materialCategory) => materialCategory.categoryId === categoryId)
+      .map((materialCategory) => materialCategory.materialId);
+
+    const filteredMaterials = db.materials.filter((material) =>
+      filteredMaterialIdsByCategory.includes(material.id)
+    );
+
+    res.send(filteredMaterials);
+  } catch (error) {
+    console.log(error);
+    res.send("Error in get one material by category");
+  }
+});
+
+// Update one material
+app.put(
+  "/materials/:id",
+  (req: Request<{ id: string }, {}, MaterialUpdate, {}>, res) => {
+    try {
+      const body: unknown = req.body;
+      const { id } = req.params;
+
+      const parsedBody = materialUpdateSchema.parse(body);
+
+      const { categoryIds, ...restParsedBody } = parsedBody;
+
+      const updatedMaterials = db.materials.map((material) => {
```

Comment:
Similar to line 194, once you find the existing material, it would be more efficient to mutate it in place.